abbyy_to_epub3/parse_abbyy.py

```diff
@@ -203,286 +269,239 @@ def parse_abbyy(self):
-            ):
-                    self.paragraphs[id]['fontstyle'] = self.fontStyles[
-                        'mainFontStyleId'
-                    ]
+            self.paragraphs[elem.get("id")] = dict(elem.attrib)
+            fontstyles = elem.iterchildren()
+            for fontstyle in fontstyles:
+                self.fontStyles[fontstyle.get("id")] = dict(fontstyle.attrib)
+
+    def process_pages(self, elem):
+        """
+        Iteratively process pages from the ABBYY file. We have to process now
+        rather than copying the pages for later processing, because deepcopying
+        an lxml element replicates the entire tree.
+        The ABBYY seems to be sometimes inconsistent about whether these
+        elements have a namespace, so be forgiving.
+        """
 
-        elif (
+        if (
             elem.tag == "{{{}}}page".format(self.ns) or
             elem.tag == "page"
         ):
-            self.pages.append(deepcopy(elem))
-        else:
-            return
+            d = {'page_no': self.page_no}
 
-        # only garbage collect if we have found & processed a node
-        elem.clear()
-
-    def parse_content(self):
-        """ Parse each page of the book.  """
-        page_no = 0
-        d = {'page_no': page_no}
-
-        self.pages[0].clear()    # clear the memory first
-        for page in self.pages:
-            pagewidth = page.get('width')
-            pageheight = page.get('height')
-            block_per_page = page.getchildren()
+            self.pagewidth = elem.get('width')
+            self.pageheight = elem.get('height')
+            block_per_page = elem.iterchildren()
             if not block_per_page:
-                page_no += 1
-                continue
-            newpage = True
+                self.page_no += 1
+                return
+            self.newpage = True
 
+            # Most pages have multiple `<block>` elements
             for block in block_per_page:
-                blockattr = block.attrib
-                blockattr['pagewidth'] = pagewidth
-                blockattr['pageheight'] = pageheight
-                if self.is_block_type(block, "Text"):
-                    paras = block.findall(".//a:par", namespaces=self.nsm)
-                    # Some blocks can have multiple styles in them. We'll treat
-                    # those as multiple blocks.
-                    for para in paras:
-                        # Get the paragraph style and text
-                        para_id = para.get("style")
-                        if para_id not in self.paragraphs:
-                            self.logger.info(
-                                'Block {} has no paragraphStyle'.format(
-                                    para_id
-                                )
-                            )
-                            self.paragraphs[para_id] = dict()
-                        text = gettext(para).strip()
+                self.parse_block(block)
```

Comment:
Do we want to clear the `block` at the end of this method?

Reply:
The blocks still get used after this so they're still needed.